### PR TITLE
(BIDS-2250) remove last attestation slot from the exporter process

### PR DESCRIPTION
**Pull request by peterbitfly**
this is a cleanup PR that removes writing the last attestation slot to the postgres db and will significantly improve the epoch export speed

includes a one time migration script that writes the last attestation slot that is currently in postgres to bigtable. this can safely be done for active validators as bigtable will only keep the most recent last attestation slot. the misc tool needs to be run with the `migrate-last-attestation-slot-bigtable` command.

### <samp>🤖 Generated by Copilot at 5189fe6</samp>

This pull request improves the efficiency and reliability of the validator attestation queries by using Bigtable as a cache layer. It refactors the code in `./db/db.go` to replace the PostgreSQL queries with Bigtable calls.

### CODE REVIEW

File: cmd/misc/main.go [hunk 1]
```diff
@@ -166,11 +166,36 @@ func main() {
...
+	validators := []types.Validator{}
+
+	err := db.WriterDb.Select(&validators, "SELECT validatorindex, lastattestationslot FROM validators WHERE lastattestationslot IS NOT NULL ORDER BY validatorindex")
+
+	if err != nil {
+		utils.LogFatal(err, "error retrieving last attestation slot", 0)
+	}
+
+	for _, validator := range validators {
+		logrus.Infof("setting last attestation slot %v for validator %v", validator.Index, validator.LastAttestationSlot)
```
Comment (D13ce) [inline cmd/misc/main.go:190]: Wrong order of the parameters.

File: db/db.go
Comment (D13ce) [inline db/db.go:1]: Now that `lastattestationslot` is not used anymore, there should be a migration file that deletes the column from the according table. Said file cannot be part of this PR of course (as it is required for the migration script), but we should not forget to add it at some point.

File: db/db.go [hunk 2]
```diff
@@ -1053,18 +1053,23 @@ func saveValidators(data *types.EpochData, tx *sqlx.Tx, client rpc.Client) error
 	}
 
 	var currentState []*types.Validator
-	err := tx.Select(&currentState, "SELECT validatorindex, withdrawableepoch, withdrawalcredentials, slashed, activationeligibilityepoch, activationepoch, exitepoch, lastattestationslot, status FROM validators;")
+	err := tx.Select(&currentState, "SELECT validatorindex, withdrawableepoch, withdrawalcredentials, slashed, activationeligibilityepoch, activationepoch, exitepoch, status FROM validators;")
```
Comment (D13ce) [inline db/db.go:1056]: Now that `lastattestationslot` is not part of the table anymore, consider removing the `db` tag and changing the datatype here:
https://github.com/gobitfly/eth2-beaconchain-explorer/blob/36a95eb96c69036c42bc6891d98cd5d177b592af/types/exporter.go#L87

**Review by D13ce — CHANGES_REQUESTED** [review]

File: db/bigtable.go [hunk 3]
```diff
@@ -612,6 +612,20 @@ func (bigtable *Bigtable) SaveAttestations(blocks map[uint64]map[string]*types.B
...
+	mutLastAttestationSlot := gcp_bigtable.NewMutation()
+	mutLastAttestationSlot.Set(ATTESTATIONS_FAMILY, fmt.Sprintf("%d", validator), gcp_bigtable.Timestamp(lastAttestationSlot*1000), []byte{})
+	err := bigtable.tableValidators.Apply(ctx, fmt.Sprintf("%s:lastAttestationSlot", bigtable.chainId), mutLastAttestationSlot)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
```
[the remark this answers is not in the document]
Comment (peterbitfly) [reply db/bigtable.go:628]: currently `tableValidators` is a private field of the `bigtable` struct. as it is now, its not possible to access that field outside of the db package. hence the function. I've added a comment that this method is not for production use.

File: db/db.go (hunk 2 above)
Comment (peterbitfly) [reply db/db.go:1056]: that needs to be done in a separate bids, as the migration script still uses that mapping

File: db/db.go
Comment (peterbitfly) [reply db/db.go:1]: ok, should be done in a separate bids

File: cmd/misc/main.go (hunk 1 above)
Comment (peterbitfly) [reply cmd/misc/main.go:190]: fixed

File: db/db.go (hunk 2 above)
Comment (D13ce) [reply db/db.go:1056]: Right :facepalm:

**Pull request merged**